source/common/upstream/load_balancer_impl.h

```diff
@@ -411,17 +430,34 @@ class RoundRobinLoadBalancer : public EdfLoadBalancerBase {
     // already exists. Note that host sources will never be removed, but given how uncommon this
     // is it probably doesn't matter.
     rr_indexes_.insert({source, seed_});
+    // If the list of hosts changes, the order of picks change. Discard the
+    // index.
+    peekahead_index_ = 0;
   }
   double hostWeight(const Host& host) override { return host.weight(); }
+  HostConstSharedPtr unweightedHostPeek(const HostVector& hosts_to_use,
+                                        const HostsSource& source) const override {
+    uint64_t* peekahead_index = const_cast<uint64_t*>(&peekahead_index_);
```

Comment:
nit: prefer making peekahead_index_ mutable to const_cast (with comment) but up to you.